Clip probabilities in ``QuantumState``

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #9761 by introducing an optional `clip` argument to the probability calculations in the ``QuantumState``. This can be used to ensure the probabilities do not exceed 1, e.g. in the ``Sampler``.

### Details and comments

Another solution would be to only manually clip inside the ``Sampler``. However this might be more generally useful so maybe adding it to the ``QuantumState`` is the better solution. If we want to backport this however, we can reduce it to a change only inside the ``Sampler``, and add this new argument for the next feature release.

I didn't add any tests or reno until we discussed this is the right approach 🙂

t-imamichi commented:
I think `QuantumState.probabilities` should return probabilities in range [0, 1]. So, there is another option to always apply `np.clip` without introducing `clip` argument.

jakelishman reviewed (commented):
Adding a parameter to an abstract method that's intended to be subclassed is a breaking API change - at the very least, it'll break Aer to some degree.  It's probably safest to not add the option and just add the clipping - it'd be good to check, but I don't think it'll have too much impact on any important performance.

t-imamichi commented on qiskit/quantum_info/states/stabilizerstate.py:
```diff
@@ -366,7 +373,11 @@ def probabilities_dict(self, qargs=None, decimals=None):
             for key, value in probs.items():
-                probs[key] = round(value, decimals)
+                prob = round(value, decimals)
+                if clip:
+                    prob = np.clip(prob, a_min=0, a_max=1)
```
if `decimals` is None, `np.clip` seems not applied even if `clip=True`.

nonhermitian commented:
Can't one just re-normalize the length of the vector using `np.linalg.norm()`?

jakelishman commented:
Pauli: It's tricky for me to be 100% sure because we're getting into the differences between floating-point and real mathematics, but I think renormalisation will always make values that are slightly above 1 safe, but it'll fail for numbers that are slightly below zero.  The dynamic range of floating-point is greater close to zero, so `-eps / (1 + eps)` is still likely to be negative, not zero.

For example:
```python
>>> import sys
>>> import numpy.linalg
>>> a = numpy.array([-sys.float_info.epsilon, 1 + sys.float_info.epsilon])
>>> a / numpy.linalg.norm(a)
array([-2.22044605e-16,  1.00000000e+00])
```

(since `sys.float_info.epsilon` is the dynamic-range limit at 1, but not at zero)

jakelishman commented:
And to finish the thought: since the matrix operations that we consider to build up these statevectors are going to include matrix multiplications that have addition/subtraction in, the magnitude of error terms in each floating-point value is going to be approximately the same on an absolute scale, rather than relative to the dynamic range around that particular value.  In other words, if we see errors like `1 + eps`, we're likely to see `-eps` as well.

Cryoris commented:
@nonhermitian Yeah I was thinking about that but can we be sure that the norm calculation and division doesn't itself have roundoff errors?

Cryoris commented:
Made the clipping implicit in 0b74ba4.

nonhermitian commented:
```python
a = np.array([-sys.float_info.epsilon, 1 + sys.float_info.epsilon])
(a*a)/ numpy.linalg.norm(a)**2
```

I just do not like seeing `'clip` there.

jakelishman commented:
Although actually yeah, come to think of it, how were we ever seeing negative numbers in the _probabilities_?  I was thinking of amplitudes only, apparently.

Cryoris commented:
I've added a renormalization too, it's true that that is likely the more subtle correction -- and then to be certain we can still clip. I also removed clipping from the stabilizer states as it seems the probabilities are consecutively divided by 2 among the possible outcomes and can therefore not exceed [0, 1].

t-imamichi commented:
Since this PR applies clip after round, it may ignore round option.

```python
from math import sqrt
from qiskit.quantum_info import Statevector

p = sqrt(1/3)
sv = Statevector([p, p, p, 0])
print(sv.probabilities(decimals=2))
```

main `[0.33 0.33 0.33 0.  ]`

this PR `[0.33333333 0.33333333 0.33333333 0.        ]`

Cryoris commented:
Good catch! That happened due to the division by the sum of the probabilities, not due to the clipping itself. It's fixed in 6b3c30a along with a test 🙂

jakelishman reviewed (approved):
Personally I think the `decimals` option on this class is a bit weird (it's pretty unusual to do that to the floats themselves as part of a probability distribution and not as part of string formatting), but never mind.

I pushed a commit to tighten the tests up, which caught an effective testing bug because of the use of `assertDictAlmostEqual` (any keys whose value is 0 should be omitted, but the expected dict contained them), because we're looking to get 0.23.3 out today and Julien's off for the day.
